apps/admin/frontend/src/components/smartcard_modal/smartcard_modal.tsx

```diff
@@ -45,7 +46,13 @@ export function SmartcardModal(): JSX.Element | null {
           smartcardType: 'system-administrator',
         });
       let contents: JSX.Element;
-      if (auth.programmableCard.programmedUser) {
+      if (
+        auth.programmableCard.programmedUser &&
+        // Writing user data may finish before writing additional data (e.g. the election
+        // definition). Don't transition from the card programming view to the card details view
+        // until all data has been written
+        !isSmartcardProgrammingInProgress(actionStatus)
```

Comment:
I was wondering about this when I saw that it was two separate writes. Is there a way that we could hide this in within the auth lib somehow so that the card programming action is atomic from the consumer's point of view?

Reply:
Yeah this piece is interesting 🤔

While we can make this seem atomic from the auth lib interface perspective, e.g. `writeUserAndData` or `programUser`, under the hood is a different story. With our current memory cards, we do ultimately store data on the card in one continuous concatenated blob. But with Java Cards, user data and other data will live in separate data slots and require separate sequential commands to the card.

We can still make things seem atomic in the Java Card world though, by tweaking how `getCardStatus` operates. Instead of having it actually read from the card on every call, it could manage a cache of the current card data and pull from that instead of the underlying card. Of course, all the challenges with caching hold, but I've been considering this anyway since, unlike on our current memory cards, retrieving user info from Java Cards will be more involved (we'll be retrieving an x509 certificate and parsing it with openssl).

Bringing this back to the current PR, I do like the idea of a single card API method for card programming so will at least make that change

Reply:
What if `getCardStatus` returned some sort of "processing" status while writes were in progress? Where would that lead?

Reply:
That seems like a very reasonable option! If it's okay, I'm going to hold off on that for now since I'd like to explore that change in the context of the Java Card implementation

I've at least removed the logic above from the smart card modal since the joint write of user and election data with memory cards is atomic again (as of https://github.com/votingworks/vxsuite/pull/3035/commits/a17960594cc076da4221ead12e9e2a27084f84c1)

Reply:
Sounds good!